As a user, I don't want an NPE when the WRES identifies a default predicted type of single valued forecasts, but its actually extended AnA

The exception:

```
2025-05-16T12:11:31.148+0000  [pool-357-thread-1] INFO EvaluationService - The evaluation failed with the following stack trace: 
wres.pipeline.InternalWresException: Could not complete project execution
    at wres.pipeline.Evaluator.evaluate(Evaluator.java:408)
    at wres.pipeline.Evaluator.evaluate(Evaluator.java:180)
    at wres.Functions.evaluate(Functions.java:135)
    at wres.server.EvaluationService.lambda$startEvaluation$2(EvaluationService.java:695)
    at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
    at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
    at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
    at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: wres.pipeline.WresProcessingException: Encountered an error while processing evaluation '7O5cI5NDVZCsQVzw4DK9yrrx7hY': 
    at wres.pipeline.Evaluator.evaluate(Evaluator.java:855)
    at wres.pipeline.Evaluator.evaluate(Evaluator.java:379)
    ... 7 common frames omitted
Caused by: wres.io.ingesting.IngestException: An ingest task could not be completed.
    at wres.io.ingesting.SourceLoader.load(SourceLoader.java:165)
    at wres.pipeline.Evaluator.evaluate(Evaluator.java:601)
    ... 8 common frames omitted
Caused by: java.util.concurrent.CompletionException: java.lang.NullPointerException: Cannot invoke "wres.config.yaml.components.TimeInterval.minimum()" because "dates" is null
    at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:315)
    at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:320)
    at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1770)
    ... 3 common frames omitted
Caused by: java.lang.NullPointerException: Cannot invoke "wres.config.yaml.components.TimeInterval.minimum()" because "dates" is null
    at wres.reading.wrds.nwm.WrdsNwmReader.getWeekRanges(WrdsNwmReader.java:485)
    at wres.reading.wrds.nwm.WrdsNwmReader.read(WrdsNwmReader.java:261)
    at wres.reading.wrds.nwm.WrdsNwmReader.read(WrdsNwmReader.java:182)
    at wres.io.ingesting.SourceLoader.lambda$readAndIngestData$3(SourceLoader.java:395)
    at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1768)
    ... 3 common frames omitted
```

The declaration is below with information omitted. The WRES identified the predicted type of single-valued forecasts when it was actually simulation or analyses data.

Thanks,

Hank

========================

```yaml
label: Prototyping NWM EANA ABRFC

observed:
  label: USGS NWIS Streamflow Observations
  sources:
  - interface: usgs nwis
    uri: https://nwis.waterservices.usgs.gov/nwis/iv
  variable: '00060'

predicted:
  label: WRDS Extended AnA Streamflow
  sources:
  - interface: wrds nwm
    uri: https://WRDS/api/nwm3.0/v3.0/ops/analysis_assim_extend
  variable: streamflow

unit: ft3/s
valid_dates:
  minimum: '2025-02-15T00:00:00Z'
  maximum: '2025-05-14T23:00:00Z'
time_scale:
  function: maximum
  period: 24
  unit: hours
rescale_lenience: observed

feature_service:
  uri: https://WRDS/api/location/v3.0/metadata
  groups:
  - group: geo_rfc
    value: aprfc
    pool: false

sampling_uncertainty:
  sample_size: 1200
  quantiles: [0.025,0.975]

threshold_sources:
- uri: omitted/hank_edited_thresholds.csv
  feature_name_from: predicted
  operator: greater equal
  apply_to: predicted

metrics:
- sample size
- probability of detection
- false alarm ratio
- bias fraction
- coefficient of determination
- mean square error skill score
- kling gupta efficiency

duration_format: hours
output_formats:
  - pairs
  - csv2
```

A separate check could probably occur very slightly earlier at the moment a reader is instantiated, but this would save very little. Most importantly, when there is sufficient declaration to know that a wrds nwm source will be read and there are no date constrains, then you get an early declaration validation error:

```
observed: foo/bar
predicted:
   sources:
     - uri: baz/qux
       interface: wrds nwm
```

```
Caused by: wres.config.yaml.DeclarationException: Encountered 3 error(s) in the declared evaluation, which must be fixed:
    - Discovered 'predicted' data sources that have a data type of 'null' and use web services, but the 'reference_dates' and/or 'valid_dates' were incomplete or undefined, which is not allowed. Please declare a complete pair of 'reference_dates' or 'valid_dates' when acquiring time-series data from web services.
  
```

Similarly, if the data `type` is clarified, you get an even more specific error:

```
observed: foo/bar
predicted:
   sources:
     - uri: baz/qux
       interface: wrds nwm
   type: single valued forecasts
```

```
Caused by: wres.config.yaml.DeclarationException: Encountered 3 error(s) in the declared evaluation, which must be fixed:
    - Discovered 'predicted' data sources that have a data type of 'single valued forecasts' and use web services, but the 'reference_dates' were incomplete or undefined, which is not allowed. Please declare a complete pair of 'reference_dates' when acquiring single valued forecasts data from web services.

```

So I think we are doing the most we can at declaration validation time with the information provided. But now we have a more meaningful error and exception at read time when an attempt is made to acquire forecast data from the wrds nwm service without `reference_dates`.

--- a/wres-config/src/wres/config/yaml/DeclarationValidator.java
+++ b/wres-config/src/wres/config/yaml/DeclarationValidator.java
@@ -4931,7 +4931,7 @@ public class DeclarationValidator
         List<EvaluationStatusEvent> events = new ArrayList<>();
 
         String messageStart = "Discovered '";
-        String messageStartOuter = "' data sources that have a data type of '";
+        String messageStartOuter = "' data sources that have a data 'type' of '";
         String messageMiddleInner = "' and use web services, but the ";
         String messageMiddleOuter =
                 " were incomplete or undefined, which is not allowed. Please declare a complete " + "pair of ";
@@ -4961,8 +4961,8 @@ public class DeclarationValidator
         {
             EvaluationStatusEvent event = eventBuilder.setEventMessage( messageStart
                                                                         + orientation
-                                                                        + messageStartOuter
-                                                                        + "null"
+                                                                        + "' data sources that have no declared data "
+                                                                        + "'type"
                                                                         + messageMiddleInner
                                                                         + REFERENCE_DATES
                                                                         + " and/or "

--- a/wres-config/test/wres/config/yaml/DeclarationValidatorTest.java
+++ b/wres-config/test/wres/config/yaml/DeclarationValidatorTest.java
@@ -369,25 +369,25 @@ class DeclarationValidatorTest
         List<EvaluationStatusEvent> events = DeclarationValidator.validate( declaration );
 
         assertAll( () -> assertTrue( DeclarationValidatorTest.contains( events, "'observed' data sources that "
-                                                                                + "have a data type of 'observations' "
-                                                                                + "and use web services, but the "
+                                                                                + "have a data 'type' of 'observations'"
+                                                                                + " and use web services, but the "
                                                                                 + "'valid_dates' were incomplete",
                                                                         StatusLevel.ERROR ) ),
                    () -> assertTrue( DeclarationValidatorTest.contains( events, "'predicted' data sources that "
-                                                                                + "have a data type of 'ensemble "
+                                                                                + "have a data 'type' of 'ensemble "
                                                                                 + "forecasts' and use web services, "
                                                                                 + "but the 'reference_dates' were "
                                                                                 + "incomplete",
                                                                         StatusLevel.ERROR ) ),
                    () -> assertTrue( DeclarationValidatorTest.contains( events, "'baseline' data sources that "
-                                                                                + "have a data type of 'single valued "
-                                                                                + "forecasts' and use web services, "
+                                                                                + "have a data 'type' of 'single valued"
+                                                                                + " forecasts' and use web services, "
                                                                                 + "but the 'reference_dates' were "
                                                                                 + "incomplete",
                                                                         StatusLevel.ERROR ) ),
                    () -> assertTrue( DeclarationValidatorTest.contains( events, "'covariate' data sources that "
-                                                                                + "have a data type of 'observations' "
-                                                                                + "and use web services, but the "
+                                                                                + "have a data 'type' of 'observations'"
+                                                                                + " and use web services, but the "
                                                                                 + "'valid_dates' were incomplete",
                                                                         StatusLevel.ERROR ) )
         );

--- a/wres-reading/src/wres/reading/wrds/nwm/WrdsNwmReader.java
+++ b/wres-reading/src/wres/reading/wrds/nwm/WrdsNwmReader.java
@@ -111,10 +111,10 @@ public class WrdsNwmReader implements TimeSeriesReader
         {
             Pair<SSLContext, X509TrustManager> sslContext = ReaderUtilities.getSslContextForWrds();
             OkHttpClient client = WebClientUtils.defaultTimeoutHttpClient()
-                                           .newBuilder()
-                                           .sslSocketFactory( sslContext.getKey().getSocketFactory(),
-                                                              sslContext.getRight() )
-                                           .build();
+                                                .newBuilder()
+                                                .sslSocketFactory( sslContext.getKey().getSocketFactory(),
+                                                                   sslContext.getRight() )
+                                                .build();
             CUSTOM_WEB_CLIENT = new WebClient( client );
         }
         catch ( PreReadException e )
@@ -452,6 +452,17 @@ public class WrdsNwmReader implements TimeSeriesReader
         if ( !( ReaderUtilities.isWrdsNwmSource( dataSource ) ) )
         {
             throw new ReadException( "Expected a WRDS NWM data source, but got: " + dataSource + "." );
+        }
+
+        if ( DeclarationUtilities.isForecast( dataSource.getContext() )
+             && Objects.nonNull( this.getDeclaration() )
+             && Objects.isNull( this.getDeclaration()
+                                    .referenceDates() ) )
+        {
+            throw new ReadException( "Encountered a WRDS NWM forecast data source, which cannot be read without "
+                                     + "'reference_dates'. If this is not a forecast data source, please clarify the "
+                                     + "data 'type' and try again. Otherwise, please declare 'reference_dates' and try "
+                                     + "again. The data source is: " + dataSource + "." );
         }
     }
 

--- a/wres-reading/test/wres/reading/wrds/nwm/WrdsNwmReaderTest.java
+++ b/wres-reading/test/wres/reading/wrds/nwm/WrdsNwmReaderTest.java
@@ -2,6 +2,7 @@ package wres.reading.wrds.nwm;
 
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
 import static org.mockserver.model.HttpRequest.request;
 
 import java.net.URI;
@@ -10,7 +11,6 @@ import java.util.Collections;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
-import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
 import org.junit.jupiter.api.AfterEach;
@@ -348,7 +348,7 @@ class WrdsNwmReaderTest
         try ( Stream<TimeSeriesTuple> tupleStream = reader.read( fakeSource ) )
         {
             List<TimeSeries<Double>> actual = tupleStream.map( TimeSeriesTuple::getSingleValuedTimeSeries )
-                                                         .collect( Collectors.toList() );
+                                                         .toList();
 
             Geometry geometry = MessageUtilities.getGeometry( Integer.toString( NWM_FEATURE_ID ),
                                                               null,
@@ -568,5 +568,75 @@ class WrdsNwmReaderTest
                 LOGGER.debug( "Count: {}.", count );
             }
         } );
+    }
+
+    @Test
+    void testReadDoesNotThrowNullPointerExceptionWhenChunkingFeaturesWithMissingReferenceDatesForForecastType()
+    {
+        URI fakeUri = URI.create( "fake" );
+
+        Source fakeDeclarationSource = SourceBuilder.builder()
+                                                    .uri( fakeUri )
+                                                    .sourceInterface( SourceInterface.WRDS_NWM )
+                                                    .build();
+
+        Dataset dataset = DatasetBuilder.builder()
+                                        .type( DataType.SINGLE_VALUED_FORECASTS )
+                                        .sources( List.of( fakeDeclarationSource ) )
+                                        .variable( VariableBuilder.builder()
+                                                                  .name( "streamflow" )
+                                                                  .build() )
+                                        .build();
+
+        DataSource fakeSource = DataSource.of( DataSource.DataDisposition.JSON_WRDS_NWM,
+                                               fakeDeclarationSource,
+                                               dataset,
+                                               Collections.emptyList(),
+                                               fakeUri,
+                                               DatasetOrientation.RIGHT,
+                                               null );
+
+        TimeInterval validDates = TimeIntervalBuilder.builder()
+                                                     .minimum( Instant.parse( "2022-01-03T00:00:00Z" ) )
+                                                     .maximum( Instant.parse( "2022-01-23T00:00:00Z" ) )
+                                                     .build();
+
+        Set<GeometryTuple> geometries
+                = Set.of( GeometryTuple.newBuilder()
+                                       .setRight( Geometry.newBuilder()
+                                                          .setName( Integer.toString( NWM_FEATURE_ID ) ) )
+                                       .build(),
+                          GeometryTuple.newBuilder()
+                                       .setRight( Geometry.newBuilder()
+                                                          .setName( "234442421" ) )
+                                       .build() );
+        Features features = FeaturesBuilder.builder()
+                                           .geometries( geometries )
+                                           .build();
+        EvaluationDeclaration declaration = EvaluationDeclarationBuilder.builder()
+                                                                        .validDates( validDates )
+                                                                        .features( features )
+                                                                        .build();
+
+        SystemSettings systemSettings = Mockito.mock( SystemSettings.class );
+        Mockito.when( systemSettings.getMaximumWebClientThreads() )
+               .thenReturn( 6 );
+        Mockito.when( systemSettings.getPoolObjectLifespan() )
+               .thenReturn( 30_000 );
+
+        // Feature chunk size of 1, with 2 features requested
+        WrdsNwmReader reader = WrdsNwmReader.of( declaration, systemSettings, 1 );
+
+        // Expect a ReadException, not an NPE: GitHub #497
+        ReadException exception = assertThrows( ReadException.class, () -> {  // NOSONAR
+            try ( Stream<TimeSeriesTuple> tupleStream = reader.read( fakeSource ) )
+            {
+                long count = tupleStream.count();
+                LOGGER.debug( "Count: {}.", count );
+            }
+        } );
+
+        assertTrue( exception.getMessage()
+                             .contains( "Encountered a WRDS NWM forecast data source" ) );
     }
 }

